### PR TITLE
[REMOVED]

### PR DESCRIPTION
**Pull request by yf225**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29278 Make all non-input arguments to functionals part of its options
* ** #29277 Pass F::*FuncOptions instead of torch::nn::*Options to functionals **

Currently, we use `torch::nn::*Options` both as module options and functional options. However, this makes it very hard to manage the parameters in `torch::nn::*Options`, because a module's constructor can take a different set of arguments than the module's equivalent functional (e.g. `torch.nn.BatchNorm1d` takes `num_features, eps=1e-5, momentum=0.1, affine=True,
track_running_stats=True`, while `F::batch_norm` takes `running_mean, running_var, weight=None, bias=None, training=False, momentum=0.1, eps=1e-5`).

This PR resolves the above problem by making `F::*FuncOptions` a different class from `torch::nn::*Options` when necessary (i.e. when a module's constructor takes a different set of arguments than the module's equivalent functional). In the rest of the cases where the module constructor takes the same set of arguments as the module's equivalent functional, `F::*FuncOptions` is an alias of `torch::nn::*Options`.

Differential Revision: [D18350164](https://our.internmc.facebook.com/intern/diff/D18350164)

### CODE REVIEW

**Comment by yf225** [comment]
@pbelevich Sorry that I screwed up on ghstack again and lost my stack information. I will open a new PR for this.

**Comment by yf225** [comment]
New PR is at https://github.com/pytorch/pytorch/pull/29318.
